# Missing values in the Dataset
- Missing values means while we getting a dataset it might have a few records are empty
- Their might be a different causses for the missing data
     1. Their might not recorded while recording
     2. Their might be a data corruption when storing the data
     3. Their might be sensor issue while getting a data
     4. they didnt provided the data whilerecording

## Types of Missing data in a Dataset
- ### Missing completely At random
    - in this the probability of data missing is random for all its records 
    - there is no pattern between other data
    - that missing data is independent of other data there is no pattern
    - it will happen in a rare cases
        - Example :-
            -  Example, suppose in a library there are some overdue books. Some values of overdue books in the computer system are missing. The reason might be a human error like the librarian forgot to type in the values. So, the missing values of overdue books are not related to any other variable/data in the system.
            
- ### Missing at random
    - in this type there might be a pattern between the missing values and the other data
    - in this there is a pattern between the missing values because
        - Example :- 
            - Suppose a poll is taken for overdue books of a library. Gender and the number of overdue books are asked in the poll. Assume that most of the females answer the poll and men are less likely to answer. So why the data is missing can be explained by another factor that is gender.
            


- ### Missing not at random
    - if the missing data is not fall under Missing at random and Missing not at random that data will fall under the Missing Not at random 
    - Missing values depend on the unobserved data
    - if there is some pattern between the data and missing data thir might be a unobserved data that can not be explained
        - Example :- 
            - suppose the name and the number of overdue books are asked in the poll for a library. So most of the people 

## How to handle the missing values in a dataset
- To handle the missing value in a dataset we can use multiple methods 

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [19]:
df=pd.read_csv('https://raw.githubusercontent.com/sahutkarsh/loan-prediction-analytics-vidhya/master/train.csv')

In [27]:
#to get if there isnulls inthe data weill use 
df.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

- As we seen above we found so many null values in the columns 
- total of 6 columns

In [10]:
df.isnull().sum().sum() # this will getthe total number of null values in the dataSet

149

As it showing we had total 149 null values in the dataset

## Figuring out the way to handle these null values in the dataset
- Analyze the each column with the missing values and figure out why the msissing values occoured 
- and we had to use a method to use to handle these missing values
- And also this is the crucial part of building the ML model
    - There are Two methods to handle the missing values
          1. Deleting the missing values
          2. Imputing the missing values with a clossest or different methods to caluclate these values and imputing then in that place


- ### Deleting the missing values
    - Generaly this is not the best way to handle these missing values as it reduces the precession of the model
    - if the missing values if types of the Missing not at random it should notbe deleted because it will effect on the dataset
    - If the missing value is of type Missing At Random or Missing Completely At Random then it can be deleted 
        - ##### The disadvantage of this method is one might end up deleting some useful data from the dataset
    - #### there are Two Different ways to delete the data
        1. Deleting the entair row
            - If the Row has manymissing values we can choose to delete that row 
            - if every row has some missing values we are endup deleting so many rows 
            - Code :- 
               - `df = train_df.dropna(axis=0)`
               - `df.isnull().sum()`

In [22]:
new_Rdf = df.dropna(axis=0) # Deleting the entair row

In [23]:
new_Rdf.isnull().sum() # To get the total number of missing values in the each columns

Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

In [24]:
new_df.shape # Shape of New Data Frame

(480, 13)

2. Deleting the entair column
    - if Certain columns have so many missing values we have to delete that column
    - Code :- 
        - `df = train_df.drop(['Dependents'],axis=1)`
        - `df.isnull().sum()`

In [30]:
new_Cdf = df.drop(['Dependents'],axis=1) # Deleting the entair Column of 'Dependents'

In [31]:
new_Cdf.isnull().sum()# To get the total number of missing values in the each columns

Loan_ID               0
Gender               13
Married               3
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

- ### 2. Imputing the missing values
    - There are different ways to impute the values inside the missing value possition
    - There are
        - #### For Numaricals features
            - 1. Replacing With Arbitrary Value
            - 2. Replacing With Mean
                - Mean is the average of the all data
            - 3. Replacing With Mode
                * Mode is the most repeated element in the dataset
            - 4. Replacing With Median
                * Median is the middle element when the data was sorted to assending order
            - 5. Replacing with previous value – Forward fill
            - 6. Replacing with next value – Backward fill
            - 7. Interpolation
         - Imputing Missing Values For Categorical Features
         - There are two ways to impute missing values for categorical features as follows:
             - Impute the Most Frequent Value
             - Impute the Value “missing”, which treats it as a Separate Category

- ### 1. Replacing With Arbitrary Value
    - Arbitrary - based on personal choice or experience
    - we replace thes values with a educated guess about the missing values as a arbitrary guess
         - Example Code
             - `df['Dependents'] = train_df['Dependents'].fillna(0)` 
             - we are replacing the null values with a zeros using fillna(0)

In [37]:
df['Dependents'] = df['Dependents'].fillna(0)# we are replacing the null values with a zeros using fillna(0)
df['Dependents'].isnull().sum() # Checking the missing values inside of the Dependents column

0

- ### 2. Replacing With Mean
    - This is the most common method for filling the missing vlues of a numiric columns 
    -  If there are outliers then the mean will not be appropriate. In such cases, outliers need to be treated first. `Learn this `
    - We ar using this method to fill the missing values in the ‘LoanAmount’ and ‘Credit_History’ columns With a mean of that column
        - Code :- 
            - `df['LoanAmount'] = df['LoanAmount'].fillna(df['LoanAmount'].mean())`
            - `df['Credit_History'] = df[‘Credit_History'].fillna(df['Credit_History'].mean())`

In [39]:
df['LoanAmount'] = df['LoanAmount'].fillna(df['LoanAmount'].mean())
df['Credit_History'] = df['Credit_History'].fillna(df['Credit_History'].mean())

In [40]:
df.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents            0
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount            0
Loan_Amount_Term     14
Credit_History        0
Property_Area         0
Loan_Status           0
dtype: int64

- As We see the Columns 'LoanAmount' and 'Credit_History' columns missing values has been handled Sucsfully and clereed

- ### 3. Replacing With Mode
    - Mode is the most frequent occouring value it is used incase of Categorical Features
        - Code :-
            - `df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])` Te replce the values in Gender column
            - `df['Married'] = df['Married'].fillna(df['Married'].mode()[0])` Te replce the values in Maried column
            - `df['Self_Employed'] = df[‘Self_Employed'].fillna(df['Self_Employed'].mode()[0])` Te replce the values in Self_employed column

In [43]:
df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])
df['Married'] = df['Married'].fillna(df['Married'].mode()[0])
df['Self_Employed'] = df['Self_Employed'].fillna(df['Self_Employed'].mode()[0])

In [44]:
df.isnull().sum()

Loan_ID               0
Gender                0
Married               0
Dependents            0
Education             0
Self_Employed         0
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount            0
Loan_Amount_Term     14
Credit_History        0
Property_Area         0
Loan_Status           0
dtype: int64

As We see the Columns 'Gender','Married' and 'Self_Employed' columns missing values has been handled Sucsfully and clereed

- ### 4. Replacing With Median
    - Median is the middlemost value. It’s better to use the median value for imputation in the case of outliers.
        - Code :- 
            - `df['Loan_Amount_Term']= df['Loan_Amount_Term'].fillna(df['Loan_Amount_Term'].median())`

In [45]:
df['Loan_Amount_Term']= df['Loan_Amount_Term'].fillna(df['Loan_Amount_Term'].median())

In [46]:
df.isnull().sum()

Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

- As We see the Columns 'Loan_Amount_Term' columns missing values has been handled Sucsfully and clereed
- And also we cleared all the Missing values in the dataset With Appropriate Methods for values

- ### 5. Replacing with previous value – Forward fill
    - In some cases, imputing the values with the previous value instead of mean, mode or median is more appropriate. This is called forward fill. 
    - It is mostly used in time series data.
        - You can use ‘fillna’ function with the parameter ‘method = ffill’
        - code 
            - `fillna(method='ffill')`
            -The ffill will fill the missing value with Previous value of that column

In [47]:
import pandas as pd
import numpy as np
test = pd.Series(range(6))
test.loc[2:4] = np.nan
test

0    0.0
1    1.0
2    NaN
3    NaN
4    NaN
5    5.0
dtype: float64

In [49]:
# Forward-Fill
test.fillna(method='ffill')

0    0.0
1    1.0
2    1.0
3    1.0
4    1.0
5    5.0
dtype: float64

- As the whole missing values were replaced by the prvious value in that column

- ### 6. Replacing with next value – Backward fill
- In backward fill, the missing value is imputed using the next value.
    - Code :- 
        - `fillna(method=‘bfill')`
        - The bfill will fill the missing value with next value of that column

In [51]:
test.fillna(method='bfill')

0    0.0
1    1.0
2    5.0
3    5.0
4    5.0
5    5.0
dtype: float64

- As the whole missing values were replaced by the next value in that column

- ### 7. Interpolation
    - Missing values can also be imputed using interpolation
    - Pandas has interpolate method can be used to replace the missing values with different interpolation methods like ‘polynomial’, ‘linear’, ‘quadratic’
    - The Default method in pandas is ‘linear’.
        - Code :-
            - `test.interpolate()`

In [52]:
test.interpolate()

0    0.0
1    1.0
2    2.0
3    3.0
4    4.0
5    5.0
dtype: float64

- Imputing Missing Values For Categorical Features
- There are two ways to impute missing values for categorical features as follows:
     - Impute the Most Frequent Value
     - Impute the Value “missing”, which treats it as a Separate Category

- ### 1. Impute the Most Frequent Value
    - We will make use of ‘SimpleImputer’ in this case and as this is a non-numeric column we can’t use mean or median but we can use most frequent value and constant.
    

In [57]:
import pandas as pd
import numpy as np
X = pd.DataFrame({'Shape':['square', 'square', 'oval', 'circle', np.nan]})
X

Shape
0  square
1  square
2    oval
3  circle
4     NaN

In [64]:
from sklearn.impute import SimpleImputer # We are importing the SimpleImputer from the sklearn.impute to impute the values

In [68]:
imputer = SimpleImputer(strategy='most_frequent') # We specified the stratagy as Most_frequent
imputer.fit_transform(X)# we are applying the stratagy to the our dataframe

array([['square'],
       ['square'],
       ['oval'],
       ['circle'],
       ['square']], dtype=object)

- AS the squre is repeated two time before so the most frequent value was added here

In [70]:
# Impute the Value “missing”, which treats it as a Separate Category
imputer = SimpleImputer(strategy='constant', fill_value='missing')
imputer.fit_transform(X)

array([['square'],
       ['square'],
       ['oval'],
       ['circle'],
       ['missing']], dtype=object)

## Replacing missing values with the Si-kit library
- Univariate Approach
    - In a Univariate approach, only a single feature is taken into consideration. You can use the class SimpleImputer and replace the missing values with mean, mode, median or some constant value.

In [73]:
import numpy as np
from sklearn.impute import SimpleImputer
imp = SimpleImputer(missing_values=np.nan, strategy='mean')
imp.fit([[1, 2], [np.nan, 3], [7, 6]])

SimpleImputer()

In [74]:
X = [[np.nan, 2], [6, np.nan], [7, 6]]
print(imp.transform(X))

[[4.         2.        ]
 [6.         3.66666667]
 [7.         6.        ]]


- Multivariate Approach
    - In a multivariate approach, more than one feature is taken into consideration. There are two ways to impute missing values considering the multivariate approach. Using KNNImputer or IterativeImputer classes.

In [75]:
import pandas as pd
df = pd.read_csv('http://bit.ly/kaggletrain', nrows=6)
cols = ['SibSp', 'Fare', 'Age']
X = df[cols]
X

SibSp     Fare   Age
0      1   7.2500  22.0
1      1  71.2833  38.0
2      0   7.9250  26.0
3      1  53.1000  35.0
4      0   8.0500  35.0
5      0   8.4583   NaN

In [76]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
impute_it = IterativeImputer()
impute_it.fit_transform(X)

array([[ 1.        ,  7.25      , 22.        ],
       [ 1.        , 71.2833    , 38.        ],
       [ 0.        ,  7.925     , 26.        ],
       [ 1.        , 53.1       , 35.        ],
       [ 0.        ,  8.05      , 35.        ],
       [ 0.        ,  8.4583    , 28.50639495]])

### Nearest Neighbors Imputations (KNNImputer)
- Missing values are imputed using the k-Nearest Neighbors approach where a Euclidean distance is used to find the nearest neighbors.

In [77]:
from sklearn.impute import KNNImputer
impute_knn = KNNImputer(n_neighbors=2)
impute_knn.fit_transform(X)

array([[ 1.    ,  7.25  , 22.    ],
       [ 1.    , 71.2833, 38.    ],
       [ 0.    ,  7.925 , 26.    ],
       [ 1.    , 53.1   , 35.    ],
       [ 0.    ,  8.05  , 35.    ],
       [ 0.    ,  8.4583, 30.5   ]])

In [78]:
import pandas as pd
import numpy as np
X = pd.DataFrame({'Age':[20, 30, 10, np.nan, 10]})
X

Age
0  20.0
1  30.0
2  10.0
3   NaN
4  10.0

In [80]:
from sklearn.impute import SimpleImputer
# impute the mean
imputer = SimpleImputer()
imputer.fit_transform(X)

array([[20. ],
       [30. ],
       [10. ],
       [17.5],
       [10. ]])

In [81]:
imputer = SimpleImputer(add_indicator=True)
imputer.fit_transform(X)

array([[20. ,  0. ],
       [30. ,  0. ],
       [10. ,  0. ],
       [17.5,  1. ],
       [10. ,  0. ]])